### ISSUE TITLE
[FR] Streamline alias code

### ISSUE THREAD

**Issue by erik-wramner** [docker-mailserver/docker-mailserver#1721]
In a recent issue we discussed how the alias file is created and then updated and how the startup script and change detector modifies not only the alias file in the container, but also the mounted configuration files for accounts and aliases. We should find a cleaner solution where the mounted files can be read only and where system accounts can be added if not already present.

The new solution should ideally allow the mounted files to be mounted read only (though of course they need to be read-write when accessed by the setup script). We could also look into the locking that is done in the start script and change detector script to avoid concurrent access - that needs to be taken into account. It is not good enough to handle concurrent access inside the container, the mounted files can also be changed outside the container. The current lock tries to handle that.

**Comment by aendeavor** [comment]
I'm completely with you here. We'll first of all need a plan, then we can implement it. Sadly, I'm not using the aliases feature, but I can be of help implementing it in the current Bash-code-base.

You also mentioned the scripts have some to-do  - I too think this should be done. I'll take a look after Christmas.

**Comment by aendeavor** [comment]
Had a quick look - the proposed code changes and the ToDo's align and overlap. You cannot touch one without the other one, so we will need to figure out the a plan first, look at the scripts and then try.

@erik-wramner if you tell me what you think is best, I can go ahead and try a first implementation.

**Comment by MakerMatrix** [comment]
My first thoughts on how to improve alias management in the start-mailserver and check-for-changes scripts are:

1) There should be the built-in idea of "system-managed" aliases, starting from those outlined in #1709 - whose purpose is to route communications about various system events/errors produced from the local accounts, to postmaster@host.tld.
2) /etc/aliases should not get overwritten from either script.  That's confusing.
3) Instead there should be something like an external function that gets imported to both scripts (maintained in one place) which checks /etc/aliases for the "system-managed" aliases and for the list (if any) in /tmp/docker-mailserver/postfix-aliases.cf, appending any of those which don't already exist.  Otherwise, don't touch.

The above has the advantage that it cannot touch any aliases that make it in by another method (e.g. editing the container directly, or user-patches.sh).

**Comment by aendeavor** [comment]
This approach seems reasonable

**Comment by erik-wramner** [comment]
Indeed. Also, it would be nice not to change the original files, allowing them to be mounted read only. We could instead make a copy and modify the copy when we normalize the lines or whatever it is all those sed calls do. We can still use the original file for the checksum.

The thing I don't know is if it is possible to take a POSIX lock on a read only file? The current code tries to lock the account file in a way that makes it impossible for other containers (such as a setup script) to change the files while they are being processed. We had several bugs with that, so I really want to keep the lock.

**Comment by aendeavor** [comment]
Alright. Before we go about coding this, I'd like to know what shall be mounted read-only specially. As the only option I know is to mount the whole volume `:ro`, this could have side-effects yet unknown to us. I'm with you all in _not_ altering `/etc/aliases` to provide a default implementation and to not override it as well. The `postfix-aliases.cf` I'm currently not sure about what is to be done. @erik-wramner can you clarify or specify what you mean exactly?

I'm too go keeping the lock:)

**Comment by erik-wramner** [comment]
Sorry for being so slow with this; I have postponed it as I don't have the time to make a proper answer. However, a bad answer is better than no answer.

check-for-changes.sh wraps the updates in a lock on postfix-accounts.cf with flock. In the same way addmailuser and the other scripts invoked from setup.sh take a POSIX lock with flock on the same file. That prevents the account file from being changed in the middle of a change detector update and ensures that only one process at a time can change the files (even if they are running in different docker containers on the host).

The implementation is far from perfect. It is still possible to change the files with an editor. The lock only helps if one is actively looking for it, as the setup script does. It is also taken on one file only (the accounts file) but there are several files involved. Still, it is the best we have now.

Regardless of how we mount the files (read/write or read only) we need to make sure that the change detector cannot run when another process is updating the files and that two separate processes cannot update the files at the same time. A POSIX lock with flock taken on a file that all containers can access guarantees that. I'm not sure but I doubt you can flock a read-only file (might work though!).

As for what we should mount read only I would ideally prefer to mount all the config files read only and work on copies if possible. They need to be mounted read/write for the setup tool though. It is probably not a big deal.

**Comment by polarathene** [comment]
Re-opening, this should be tackled at some point but is low priority due to other work queued up that should come first. I don't expect this to be worked on until Feb 2022 or later.

Notes:

- @MakerMatrix outlined what needs to be done in [their comment above](https://github.com/docker-mailserver/docker-mailserver/issues/1721#issuecomment-748740921).
- The unification step between startup and `check-for-changes.sh` has been handled with the recent `v10.3.0` release. There is more cleanup and refactoring due there before improvements such as this issue are tacklled however.
- @erik-wramner provided a nice overview here regarding [file locking usage with `flock`](https://github.com/docker-mailserver/docker-mailserver/issues/1721#issuecomment-756714344) (cheers!) which has unfortunately regressed since ([as detailed by me here](https://github.com/docker-mailserver/docker-mailserver/pull/2157#pullrequestreview-817286005)) and needs to be brought back once the compatibility issue that was experienced with NFS is resolved (_the reason `flock` was replaced mid 2021_).
- Erik also introduced the original `flock` support back in mid 2019.
- Presently file locking is AFAIK, now locking to the scripts running the operation, rather than the actual config files that were previously locked against.
